homeassistant/components/recorder/statistics.py

```diff
@@ -1088,32 +1088,39 @@ def reduce_day_ts_factory() -> (
     ]
 ):
     """Return functions to match same day and day start end."""
+    _boundries: tuple[float, float] = (0, 0)
+
     # We have to recreate _local_from_timestamp in the closure in case the timezone changes
     _local_from_timestamp = partial(
         datetime.fromtimestamp, tz=dt_util.DEFAULT_TIME_ZONE
     )
-    # We create _as_local_cached in the closure in case the timezone changes
-    _as_local_cached = lru_cache(maxsize=6)(_local_from_timestamp)
-
-    def _as_local_date(time: float) -> date:
-        """Return the local date of a datetime."""
-        return _local_from_timestamp(time).date()
-
-    _as_local_date_cached = lru_cache(maxsize=6)(_as_local_date)
 
     def _same_day_ts(time1: float, time2: float) -> bool:
         """Return True if time1 and time2 are in the same date."""
-        return _as_local_date_cached(time1) == _as_local_date_cached(time2)
+        nonlocal _boundries
+        if not _boundries[0] <= time1 < _boundries[1]:
+            _boundries = _day_start_end_ts_cached(time1)
+        return _boundries[0] <= time2 < _boundries[1]
 
     def _day_start_end_ts(time: float) -> tuple[float, float]:
         """Return the start and end of the period (day) time is within."""
-        start = dt_util.as_utc(
-            _as_local_cached(time).replace(hour=0, minute=0, second=0, microsecond=0)
+        start_local = _local_from_timestamp(time).replace(
+            hour=0, minute=0, second=0, microsecond=0
+        )
+        # We need to add 4 hours to the end to make sure we are in the next week
+        # before 0ing out the the time in case there is a DST change
```

Comment:
Adding 4 hours may be overkill, we may just need to do days. Need to check how this is handled over a DST boundry

Reply:
The four hours isn't needed
```
>>> datetime.datetime(year=2022,month=3,day=13,hour=0,minute=0,second=0,tzinfo=zoneinfo.ZoneInfo(key='america/chicago')) + datetime.timedelta(days=7)
datetime.datetime(2022, 3, 20, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='america/chicago'))
>>> datetime.datetime(year=2022,month=11,day=6,hour=0,minute=0,second=0,tzinfo=zoneinfo.ZoneInfo(key='america/chicago')) + datetime.timedelta(days=7)
datetime.datetime(2022, 11, 13, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='america/chicago'))
```